### PR TITLE
Fix invalid conversions between ECI coordinates

### PR DESCRIPTION
**Pull request by parzivail**
There exist some invalid conversions when observing ECI coordinates at different times, which this PR addresses. The underlying problem may be the conversion from one ECI to another given only time difference between them.

### CODE REVIEW

File: SGP.NET/SGP.NET.csproj [hunk 1]
```diff
@@ -25,6 +25,10 @@ Remove unused time parameter in Coordinate's GetFootprint,
 Made constants constant</PackageReleaseNotes>
   </PropertyGroup>
 
+  <PropertyGroup Condition="'$(Configuration)|$(TargetFramework)|$(Platform)'=='Debug|netstandard1.4|AnyCPU'">
+    <DocumentationFile>C:\Users\Administrator\source\repos\SGP.NET\SGP.NET\SGP.NET.xml</DocumentationFile>
+  </PropertyGroup>
+
```
Comment (bretcope) [inline SGP.NET/SGP.NET.csproj]: Heads up, this hard-coded absolute path will make the build fail for most people (as it did for me).

**Comment by bretcope** [comment]
I verified that this change results in accurate pass predictions. Thanks!

File: SGP.NET/SGP.NET.csproj (hunk 1 above)
Comment (parzivail) [reply SGP.NET/SGP.NET.csproj]: Strange that it auto-generated an absolute path. I'll fix it before I merge.

**Comment by parzivail** [comment]
Perfect. I'll fix the build issue and merge.

**Comment by davidtgillard** [comment]
> 
> 
> I verified that this change results in accurate pass predictions. Thanks!

Hi guys, sorry to jump in on this... @bretcope, would you mind describing how you performed the verification (inputs and result)? I'm comparing the results of Observe with some other propagators (e.g. Orekit) using this fix and still not getting the right output. It might very well be PEBCAK but I thought I better check before this gets closed.

**Comment by bretcope** [comment]
@davidtgillard I grabbed the TLE for a spacecraft that my company operates, ran pass predictions for one of the ground stations we use (via `GroundStation.Observe()`), then compared the start/end times and maximum elevation angle against a few different sources: STK, [heavens-above](https://www.heavens-above.com/), as well as the pass schedule given to us by the ground station provider. They all matched within about a second.

I didn't run extensive tests, but that same procedure produced wildly inaccurate results before this change.

In case there was any confusion, you have to check out the `fix-7` branch, comment out or delete a few lines from the SGP.NET.csproj (commented on above), and build the project yourself in order to test this change. It's not on nuget.org or the master branch at this point.

**Comment by parzivail** [comment]
A release build, 1.0.6, has been uploaded to NuGet and will be available as soon as it's indexed.

**Comment by davidtgillard** [comment]
@bretcope thanks for the prompt response.

I was doing essentially the same as you, but experiencing a timezone offset. When I dug into it again this morning, I noticed that the result would change based upon the timezone of my system clock. 

Turns out, this is because the `Round` extension in TimeExtensions.cs doesn't preserve the Kind, and so the start and end values of `SatelliteVisibilityPeriod` end up with an unspecified kind. `Round` could be easily modified, but I think the right thing to do here is to replace DateTime with DateTimeOffset (which is probably [preferred](https://docs.microsoft.com/en-us/dotnet/standard/datetime/choosing-between-datetime) for this use). Incidentally, I already did this when I was trying out the fix branch for #7...but for some reason I didn't change `Round` so I never noticed a change.

So, @parzivail, if you're happy with changing DateTime to DateTimeOffset, I'll generate a pull request for you later. (Incidentally, I'm also happy to help out with some unit testing in the future.)

**Comment by parzivail** [comment]
My apologies, I haven't gotten around to better documentation which would have saved you all of this trouble - all of the times are UTC. In light of this I think migrating to `DateTimeOffset` is a fantastic idea. 

I have about a week before my classes start so I'll also spend some time adding some unit tests, examples, and more comprehensive documentation besides a class and method list. 

Thank you for all the help!

**Comment by bretcope** [comment]
For what it's worth, I think having times consistently in UTC is much better than supporting time zones in any way. `DateTime` has a `ToUniversalTime()` method which could be called on any time parameter to ensure it's in UTC before further processing. Calling `ToUniversalTime` would also be a non-breaking change, whereas switching everything to `DateTimeOffset` is a breaking change.

In my experience, `DateTime` is still the more common datatype in .NET, probably because servers primarily deal with UTC. There are no benefits to using DateTimeOffset if your times are UTC. The `DateTimeOffset` struct is literally implemented as a [DateTime field and a 16-bit integer representing offset in minutes](https://github.com/dotnet/corefx/blob/d3911035f2ba3eb5c44310342cc1d654e42aa316/src/Common/src/CoreLib/System/DateTimeOffset.cs#L52-L53). Furthermore, DateTime is a 64-bit struct, which means it can be passed in one register on a 64-bit machine, whereas DateTimeOffset is 80 bits and has to span registers or spill onto the stack. A micro-optimization for sure, but my point is that I don't see an advantage in switching to DateTimeOffset.

If a consumer wants to convert to or from local time, they can still do that, and it's really easy for them to convert back and forth from DateTimeOffset if desired. But I would vote for the library to keep using DateTime, perhaps with some added calls to `ToUniversalTime()`, as I mentioned.

**Comment by parzivail** [comment]
I hadn't considered that it would be a breaking change. I think it should be a simple fix to add statements to each method that consumes a `DateTime` to ensure that it's UTC. I'll make an issue for this to consolidate the change effort.

**Comment by davidtgillard** [comment]
There is an implicit conversion of a [DateTime to DateTimeOffset](https://docs.microsoft.com/en-us/dotnet/api/system.datetimeoffset.op_implicit?view=netframework-4.8). So it shouldn't be breaking.

Thanks for the peek under the hood at `DateTimeOffset` - I didn't realize that. I suppose, if there were optimizations to be performed in, say, a function call with a tight loop, one could use `DateTimeOffset` in the interface and use `DateTime` under the hood (although I must confess I don't need that optimization for any of my applications).

Part of the problem folks have with `DateTime` is that the behaviour of `ToUniversalTime()` depends upon the Kind. When calling `ToUniversalTime()`, if  the Kind is UTC, it will behave as expected. However, if Kind is Unspecified, then it is assumed the datetime is meant to be local, and the conversion to UTC is performed. Which is fine if you know that, but the behavior I was experiencing was that - I was passing in UTC DateTimes but getting back DateTimes with a Kind of Unspecified. 
 
If the intent of the library is to always work in UTC, I think it's better to make this explicit in the interface to use `DateTimeOffset`. But it's not my lib, and I'll support any decision provided it's consistent.

**Comment by davidtgillard** [comment]
Wait a minute. I'm sorry, I was wrong - it _is_ breaking. Implicit conversion in one direction does not full backwards-compatibility make.

**Comment by bretcope** [comment]
I'm not 100% against changing the inputs to use DateTimeOffset, but it may not solve the problem to the extent you'd like. DateTime is still used more common, and if someone generates an unspecified DateTime, you'll get exactly the same problem as `ToUniversalTime` during the implicit cast which will simply mask the problem.

To truly prevent this problem, I think you'd need to use DateTime and throw an exception when Kind is Unspecified. That would force the caller to fix their code rather than get wrong results. I don't see DateTimeOffset giving that same guarantee.

Changing _outputs_ (altering data structures) is a breaking change because there is no implicit cast from DateTime to DateTimeOffset, and DateTime is the normal data structure used to interact with things like SQL and lots of other systems.

**Comment by parzivail** [comment]
I think it could be a good compromise to throw an exception when a `DateTime` is provided in an undefined state, it would definitely make some bugs easier to see and would reinforce knowing exactly what times you're requesting propagation about.

**Comment by davidtgillard** [comment]
Would the Kind of output DateTimes be kept consistent with input DateTimes? Or would the Kind of output DateTimes always be UTC?

**Comment by bretcope** [comment]
@davidtgillard perhaps this discussion should be moved to #9. I'll respond there.

**Pull request merged**